### PR TITLE
[WIP] reduce size of Loc struct

### PR DESCRIPTION
**Pull request by benjones**
Wanted to get feedback on this approach to shrinking the size of the Loc struct in globals by storing an index into a global array of filenames.  Some other things I've thought about but aren't here are:

* using __gshared/thread safety?
* Searching for repeats when setting the filename?  
* I didn't use the Array class because it can't store slices.
* The use of a mutable global removes purity from most member functions... Is this a good tradeoff?

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
I don't have a good solution but it's not ideal to rely on global state. We're trying to remove global state.

**Comment by wilzbach** [comment]
For DMD as a library it would be crucial to be able to reset the table as the DMD server needs to run for a prolonged amount of time.

File: src/dmd/globals.d [hunk 1]
```diff
@@ -532,34 +545,69 @@ nothrow:
...
+        setFilename(fname);
+    }
+    
+    @property extern(C++) void filename(const(char)* fname) nothrow
+    {
+        import dmd.utils : toDString;
+        setFilename(fname.toDString);
+    }
+    
+    private extern(D) void setFilename(const char[] fname) nothrow
```
Comment (jacob-carlborg) [inline src/dmd/globals.d]: Why this extra method? Why not do this directly in the property?

File: src/dmd/globals.d [hunk 2]
```diff
@@ -478,24 +478,37 @@ alias d_uns64 = uint64_t;
 // file location
 struct Loc
 {
-    const(char)* filename; // either absolute or relative to cwd
+    uint fileIndex;
     uint linnum;
     uint charnum;
```
Comment (jacob-carlborg) [inline src/dmd/globals.d]: Can we move the line and column numbers to the table as well? Then we wouldn’t need to pass `Loc` by reference.

**Comment by jacob-carlborg** [comment]
> using __gshared/thread safety?

DMD is not thread safe, `__gshared` is what’s used. 

> I didn't use the Array class because it can't store slices.

That’s fine.

> The use of a mutable global removes purity from most member functions... Is this a good tradeoff?

No, this is very bad. I worked quite hard to make code pure. But I don’t have a good solution. Since `Loc` is used in many places not much code that otherwise could have been pure cannot be pure. Pure is the best tool we have to identify global state and enforce no new global state is introduced.

**Comment by benjones** [comment]
What about this: 

* Move the filename table stuff to a struct FileTable with methods
* `int lookup(char[] filename` and `void reset()`.
* Pass ref FileTable to the Loc constructor and filename methods
* make the `lookup` method search for duplicates.  For the equals methods, then we don't have to look to the table to check for equality.  toHash could have Loc objects built with different tables hash to the same value though

**Comment by jacob-carlborg** [comment]
> What about this:

Sure, that could work.

> Pass ref FileTable to the Loc constructor and filename methods

But you need to store a pointer to `FileTable`, or?

**Comment by benjones** [comment]
> But you need to store a pointer to `FileTable`, or?
No, you'd pass it explicitly to the methods that need it.  I'm not sure at the moment how invasive of a change it would be

**Comment by jacob-carlborg** [comment]
> No, you'd pass it explicitly to the methods that need it

Ah, I see.

> I'm not sure at the moment how invasive of a change it would be

It would be quite invasive. Even though parameters with default values are discourage in the DMD code base, I think we can do an exception for this. Then it would be possible to replace the default values with explicit values in smaller chunks. I did that with `toChars`:

https://github.com/dlang/dmd/blob/b2c4add2010f3bf0c4c015e8c57a3014fed9277a/src/dmd/globals.d#L495

**Comment by benjones** [comment]
Another direction might be turning loc structs into strings earlier in the error reporting process. At what point do we only care about locs for the purpose of a diagnostic message...

**Comment by jacob-carlborg** [comment]
I know some people would like to report diagnostics as XML or JSON. Formatting to a string early would not make that possible.

**Comment by jacob-carlborg** [comment]
Why closing this?

**Comment by benjones** [comment]
I don't have a good idea of how to move forward on it.  I'll reopen if I figure something out.

**Pull request closed without merging**